Inference crash: `Slice has no attribute 'qname'`

### Bug description

```python
x = slice(42)
x()
```

### Configuration

_No response_

### Command used

```shell
pylint asd.py
```


### Pylint output

```shell
************* Module asd
asd.py:1:0: C0114: Missing module docstring (missing-module-docstring)
Exception on node <Call l.2 at 0x1040c31f0> in file '/private/tmp/asd.py'
Traceback (most recent call last):
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 90, in walk
    callback(astroid)
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/checkers/method_args.py", line 72, in visit_call
    and inferred.qname() in self.linter.config.timeout_methods
AttributeError: 'Slice' object has no attribute 'qname'
asd.py:1:0: F0002: asd.py: Fatal error while checking 'asd.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/Users/tusharsadhwani/Library/Caches/pylint/pylint-crash-2023-01-17-14-57-43.txt'. (astroid-error)
```


### Expected behavior

No crash.

### Pylint version

```shell
pylint 2.15.10
astroid 2.13.2
Python 3.10.6 (main, Aug 30 2022, 04:58:14) [Clang 13.1.6 (clang-1316.0.21.2.5)]
```


### OS / Environment

MacOS Monterey

### Additional dependencies

_No response_

Also, weirdly enough a similar crash also happens in `RefactoringChecker` if I tweak the code a little bit:

### Code
```python
class Foo:
    def __init__(self, foo=slice(42)):
        self.foo = foo


def bar():
    i = Foo()
    i.foo()
    return 100
```

### Output

```console
$ pylint /tmp/asd.py
************* Module asd
/tmp/asd.py:1:0: C0114: Missing module docstring (missing-module-docstring)
/tmp/asd.py:1:0: C0115: Missing class docstring (missing-class-docstring)
/tmp/asd.py:3:8: C0104: Disallowed name "foo" (disallowed-name)
/tmp/asd.py:2:23: C0104: Disallowed name "foo" (disallowed-name)
/tmp/asd.py:1:0: R0903: Too few public methods (0/2) (too-few-public-methods)
/tmp/asd.py:6:0: C0116: Missing function or method docstring (missing-function-docstring)
/tmp/asd.py:6:0: C0104: Disallowed name "bar" (disallowed-name)
Exception on node <FunctionDef.bar l.6 at 0x1025149a0> in file '/tmp/asd.py'
Traceback (most recent call last):
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 95, in walk
    callback(astroid)
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 977, in leave_functiondef
    self._check_consistent_returns(node)
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 1859, in _check_consistent_returns
    ) and self._is_node_return_ended(node):
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 1971, in _is_node_return_ended
    return any(self._is_node_return_ended(_child) for _child in node.get_children())
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 1971, in <genexpr>
    return any(self._is_node_return_ended(_child) for _child in node.get_children())
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 1971, in _is_node_return_ended
    return any(self._is_node_return_ended(_child) for _child in node.get_children())
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 1971, in <genexpr>
    return any(self._is_node_return_ended(_child) for _child in node.get_children())
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 1941, in _is_node_return_ended
    if self._is_function_def_never_returning(funcdef_node):
  File "/private/tmp/venv3/lib/python3.10/site-packages/pylint/checkers/refactoring/refactoring_checker.py", line 2000, in _is_function_def_never_returning
    return node.qname() in self._never_returning_functions
AttributeError: 'Slice' object has no attribute 'qname'
/tmp/asd.py:1:0: F0002: /tmp/asd.py: Fatal error while checking '/tmp/asd.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/Users/tusharsadhwani/Library/Caches/pylint/pylint-crash-2023-01-17-15-02-58.txt'. (astroid-error)
```

--- a/doc/whatsnew/fragments/8067.bugfix
+++ b/doc/whatsnew/fragments/8067.bugfix
@@ -1,0 +1,3 @@
+Fixed a crash when inferring a value and using its qname on a slice that was being incorrectly called.
+
+Closes #8067

--- a/pylint/checkers/method_args.py
+++ b/pylint/checkers/method_args.py
@@ -8,7 +8,7 @@ from __future__ import annotations
 
 from typing import TYPE_CHECKING
 
-from astroid import arguments, nodes
+from astroid import arguments, bases, nodes
 
 from pylint.checkers import BaseChecker, utils
 from pylint.interfaces import INFERENCE
@@ -69,7 +69,9 @@ class MethodArgsChecker(BaseChecker):
         if (
             inferred
             and not call_site.has_invalid_keywords()
-            and hasattr(inferred, "qname")
+            and isinstance(
+                inferred, (nodes.FunctionDef, nodes.ClassDef, bases.UnboundMethod)
+            )
             and inferred.qname() in self.linter.config.timeout_methods
         ):
             keyword_arguments = [keyword.arg for keyword in node.keywords]

--- a/pylint/checkers/method_args.py
+++ b/pylint/checkers/method_args.py
@@ -69,6 +69,7 @@ class MethodArgsChecker(BaseChecker):
         if (
             inferred
             and not call_site.has_invalid_keywords()
+            and hasattr(inferred, "qname")
             and inferred.qname() in self.linter.config.timeout_methods
         ):
             keyword_arguments = [keyword.arg for keyword in node.keywords]

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -1611,18 +1611,17 @@ class RefactoringChecker(checkers.BaseTokenChecker):
             # the result of this call was already assigned to a variable and will be checked when leaving the scope.
             return
         inferred = utils.safe_infer(node.func)
-        if not inferred:
+        if not inferred or not isinstance(
+            inferred, (nodes.FunctionDef, nodes.ClassDef, bases.UnboundMethod)
+        ):
             return
         could_be_used_in_with = (
             # things like ``lock.acquire()``
-            hasattr(inferred, "qname")
-            and (
-                inferred.qname() in CALLS_THAT_COULD_BE_REPLACED_BY_WITH
-                or (
-                    # things like ``open("foo")`` which are not already inside a ``with`` statement
-                    inferred.qname() in CALLS_RETURNING_CONTEXT_MANAGERS
-                    and not _is_part_of_with_items(node)
-                )
+            inferred.qname() in CALLS_THAT_COULD_BE_REPLACED_BY_WITH
+            or (
+                # things like ``open("foo")`` which are not already inside a ``with`` statement
+                inferred.qname() in CALLS_RETURNING_CONTEXT_MANAGERS
+                and not _is_part_of_with_items(node)
             )
         )
         if could_be_used_in_with and not _will_be_released_automatically(node):

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -1615,11 +1615,14 @@ class RefactoringChecker(checkers.BaseTokenChecker):
             return
         could_be_used_in_with = (
             # things like ``lock.acquire()``
-            inferred.qname() in CALLS_THAT_COULD_BE_REPLACED_BY_WITH
-            or (
-                # things like ``open("foo")`` which are not already inside a ``with`` statement
-                inferred.qname() in CALLS_RETURNING_CONTEXT_MANAGERS
-                and not _is_part_of_with_items(node)
+            hasattr(inferred, "qname")
+            and (
+                inferred.qname() in CALLS_THAT_COULD_BE_REPLACED_BY_WITH
+                or (
+                    # things like ``open("foo")`` which are not already inside a ``with`` statement
+                    inferred.qname() in CALLS_RETURNING_CONTEXT_MANAGERS
+                    and not _is_part_of_with_items(node)
+                )
             )
         )
         if could_be_used_in_with and not _will_be_released_automatically(node):

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -2000,7 +2000,7 @@ class RefactoringChecker(checkers.BaseTokenChecker):
             )
         try:
             return node.qname() in self._never_returning_functions
-        except TypeError:
+        except (TypeError, AttributeError):
             return False
 
     def _check_return_at_the_end(self, node: nodes.FunctionDef) -> None:

--- a/tests/functional/r/regression_02/regression_8067.py
+++ b/tests/functional/r/regression_02/regression_8067.py
@@ -1,0 +1,4 @@
+"""Regression tests for inferred.qname missing"""
+
+x = slice(42)
+x()  # [not-callable]

--- a/tests/functional/r/regression_02/regression_8067.py
+++ b/tests/functional/r/regression_02/regression_8067.py
@@ -1,4 +1,17 @@
 """Regression tests for inferred.qname missing"""
 
+# pylint: disable=missing-docstring,too-few-public-methods,disallowed-name
+
 x = slice(42)
 x()  # [not-callable]
+
+
+class Foo:
+    def __init__(self, foo=slice(42)):
+        self.foo = foo
+
+
+def bar():
+    i = Foo()
+    i.foo()
+    return 100

--- a/tests/functional/r/regression_02/regression_8067.txt
+++ b/tests/functional/r/regression_02/regression_8067.txt
@@ -1,0 +1,1 @@
+not-callable:4:0:4:3::x is not callable:UNDEFINED

--- a/tests/functional/r/regression_02/regression_8067.txt
+++ b/tests/functional/r/regression_02/regression_8067.txt
@@ -1,1 +1,1 @@
-not-callable:4:0:4:3::x is not callable:UNDEFINED
+not-callable:6:0:6:3::x is not callable:UNDEFINED

